### PR TITLE
feat: Cylinder and Disc surface merging

### PR DESCRIPTION
**Pull request by paulgessinger**
This PR allows merging of two cylinder and disc surfaces. Merging is allowed in the $z$ and $r\times\phi$ direction for cylinders, and in the $r$ and $\phi$ direction for discs. The surfaces need to be consistent, i.e. their bounds dimensions and transforms need to line up.

- For **cylinders**:
  - $z$ merging: $z$ edges must be aligned, no relative $xy$ shift or any rotation is allowed.
  - $r\times\phi$ merging: $r\times\phi$ edges must be aligned, no $xyz$ shift or any rotation is allowed
- For **discs**:
  - $r$ merging: $r$ edges must be aligned, no $xyz$ shift or rotation is allowed
  - $\phi$ merging: $\phi$ edges must be aligned, no $xyz$ shift or rotation is allowed.

Cylinder_Merging
Disc_Merging

Blocked by:
- #3290

### CODE REVIEW

**Comment by andiwand** [comment]
As always: very nice drawings!

File: Core/src/Surfaces/CylinderSurface.cpp [hunk 1]
```diff
@@ -353,3 +362,139 @@ Acts::CylinderSurface::localCartesianToBoundLocalDerivative(
...
+
+  if (std::abs(translation[0]) > tolerance ||
+      std::abs(translation[1]) > tolerance) {
+    ACTS_ERROR(
+        "CylinderSurface::merge: surfaces have relative translation in x/y");
+    throw std::invalid_argument(
+        "CylinderSurface::merge: surfaces have relative translation in x/y");
+  }
+
+  // @TODO:Overlap checks are missing?
```
Comment (andiwand) [inline Core/src/Surfaces/CylinderSurface.cpp:429]: For now of for later?

File: Core/src/Surfaces/DiscSurface.cpp [hunk 2]
```diff
@@ -373,3 +376,120 @@ double Acts::DiscSurface::pathCorrection(const GeometryContext& gctx,
...
+  }
+
+  Vector3 translation = otherLocal.translation();
+
+  if (!translation.isApprox(Vector3::Zero())) {
+    ACTS_ERROR(
+        "DiscSurface::merge: surfaces have relative translation in x/y/z");
+    throw std::invalid_argument(
+        "DiscSurface::merge: surfaces have relative translation in x/y/z");
+  }
```
Comment (andiwand) [inline Core/src/Surfaces/DiscSurface.cpp:406]: in the `CylinderSurface` implementation we use the surface tolerance. Should we align that to use one of both?

**Review by andiwand — DISMISSED** [review]
Had a quick look - I think this is a really nice implementation!

File: Core/src/Surfaces/CylinderSurface.cpp (hunk 1 above)
Comment (paulgessinger) [reply Core/src/Surfaces/CylinderSurface.cpp:429]: The TODO is obsolete, I'll remove this

File: Core/include/Acts/Surfaces/detail/MergeHelper.hpp [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+#include "Acts/Definitions/Algebra.hpp"
+#include "Acts/Definitions/Tolerance.hpp"
+#include "Acts/Definitions/Units.hpp"
+#include "Acts/Utilities/Logger.hpp"
+#include "Acts/Utilities/detail/periodic.hpp"
+
+#include <utility>
+namespace Acts::detail {
+
+std::pair<ActsScalar, ActsScalar> mergedPhiSector(
```
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Core/include/Acts/Surfaces/detail/MergeHelper.hpp:20]: Will do.

**Comment by paulgessinger** [comment]
Should be good for follow-up @andiwand.

**Comment by andiwand** [comment]
@paulgessinger I think a couple of other changes sneaked in now

**Comment by paulgessinger** [comment]
Yeah I messed up @andiwand  😬

**Comment by paulgessinger** [comment]
Ok, should be back @andiwand, reflog saved me.

**Review by andiwand — APPROVED** [review]

File: Core/include/Acts/Utilities/Logger.hpp
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Core/include/Acts/Utilities/Logger.hpp:1]: That should be in https://github.com/acts-project/acts/pull/3290. Not sure right now why it shows in the diff here.

We need the change however, yeah.

**Review by andiwand — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :red_circle: Athena integration test results [e1a76b33e75817513e1233252d2e160522cec082]
### :red_circle: Some tests have failed!
Please investigate the [pipeline](https://gitlab.cern.ch/acts/acts-athena-ci/-/pipelines/7571261)!

|status|job|report|
|------|---|------|
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40085640)||
|:green_circle: | [test_ActsDumpGeometryIdentifiers](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40085639)||
|:green_circle: | [test_ActsCheckObjectCountsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40085638)||
|:green_circle: | [test_ActsEFTrackFit](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40085637)||
|:green_circle: | [test_ActsPersistifySeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40085636)||
|:green_circle: | [test_ActsBenchmarkWithSpot](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40085635)||
|:green_circle: | [test_ActsAnalogueClustering](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40085634)||
|:green_circle: | [test_ActsWorkflowHeavyIons](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40085633)||
|:green_circle: | [test_ActsWorkflowFastTracking](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40085632)||
|:green_circle: | [test_ActsWorkflowCached](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40085631)||
|:green_circle: | [test_ActsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40085630)||
|:green_circle: | [test_ActsValidateAmbiguityResolution](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40085629)||
|:green_circle: | [test_ActsValidateResolvedTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40085628)||
|:green_circle: | [test_ActsValidateTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40085627)||
|:green_circle: | [test_ActsValidateActsCoreSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40085626)||
|:green_circle: | [test_ActsValidateActsSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40085625)||
|:green_circle: | [test_ActsValidateSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40085624)||
|:green_circle: | [test_ActsValidateOrthogonalSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40085623)||
|:green_circle: | [test_ActsValidateClusters](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40085622)||
|:green_circle: | [test_ActsPersistifyEDM](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40085621)||
|:green_circle: | [test_ActsGSFRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40085620)||
|:green_circle: | [test_ActsKfRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40085619)||
|:green_circle: | [test_ActsExtrapolationAlgTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40085618)||
|:green_circle: | [test_ActsITkTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40085617)||
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40085616)||
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40085615)||
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40085614)||
|:red_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40085613)||
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40085612)||
|:green_circle: | [run_art_test: test_data18_13TeV_1000evt](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40085611)||
|:green_circle: | [run_art_test: test_ttbarPU40_reco](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/40085610)||

**Pull request merged**